### ISSUE TITLE
Unable to join an updated master node.

### ISSUE THREAD

**Issue by i5Js** [k3s-io/k3s#2788]
**Environmental Info:**
K3s Version: HA installation mixed version 1.19 & 1.20

k3s version v1.19.5+k3s2 (746cf403) (Master 1)
k3s version v1.20.0+k3s2 (2ea6b163) (Master 2, which doesn't works)
go version go1.15.5 (Master 2, which doesn't works)
Node(s) CPU architecture, OS, and Version: 

Linux noldork3sM1 5.4.79-v7+ #1373 SMP Mon Nov 23 13:22:33 GMT 2020 armv7l GNU/Linux
Linux noldork3sM2 5.4.79-v7+ #1373 SMP Mon Nov 23 13:22:33 GMT 2020 armv7l GNU/Linux
Linux noldork3sN1 5.4.79-v7+ #1373 SMP Mon Nov 23 13:22:33 GMT 2020 armv7l GNU/Linux (Worker node, k3s versionk3s version v1.20.0+k3s2 (2ea6b163) go version go1.15.5. Joined and working)
Cluster Configuration:

2xMaster
4xWorkers
**Describe the bug:**

After updated noldork3sM2, the node is always on NoReady status
noldork3sm2   NotReady   control-plane,master 
**Steps To Reproduce:**

- Installed K3s: 
`export K3S_KUBECONFIG_MODE="644"
export INSTALL_K3S_EXEC=" --tls-san noldork3s.noldor.local --no-deploy servicelb --disable traefik --node-taint k3s-controlplane=true:NoSchedule --datastore-endpoint mysql://user:password@tcp(server:3306)/database"
 curl -sfL https://get.k3s.io | sh `
**Expected behavior:**

The node would join to the cluster successfully
**Actual behavior:**

Stuck in NoReady
**Additional context / logs:**

```
`Jan  8 12:15:30 localhost k3s[439]: E0108 12:15:30.212199     439 leaderelection.go:325] error retrieving resource lock kube-system/kube-scheduler: Get "https://127.0.0.1:6444/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=10s": context deadline exceeded
Jan  8 12:15:36 localhost k3s[439]: E0108 12:15:36.428889     439 leaderelection.go:325] error retrieving resource lock kube-system/kube-controller-manager: Get "https://127.0.0.1:6444/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-controller-manager?timeout=10s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
Jan  8 12:15:36 localhost k3s[439]: E0108 12:15:36.707739     439 controller.go:144] failed to ensure lease exists, will retry in 7s, error: Get "https://127.0.0.1:6443/apis/coordination.k8s.io/v1/namespaces/kube-node-lease/leases/noldork3sm2?timeout=10s": context deadline exceeded
Jan  8 12:15:38 localhost k3s[439]: E0108 12:15:38.260039     439 leaderelection.go:325] error retrieving resource lock kube-system/cloud-controller-manager: the server was unable to return a response in the time allotted, but may still be processing the request (get leases.coordination.k8s.io cloud-controller-manager)
Jan  8 12:15:42 localhost k3s[439]: E0108 12:15:42.794165     439 leaderelection.go:325] error retrieving resource lock kube-system/kube-scheduler: Get "https://127.0.0.1:6444/apis/coordination.k8s.io/v1/namespaces/kube-system/leases/kube-scheduler?timeout=10s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
Jan  8 12:15:43 localhost k3s[439]: E0108 12:15:43.086538     439 storage_flowcontrol.go:137] failed creating mandatory flowcontrol settings: failed getting mandatory FlowSchema exempt due to the server was unable to return a response in the time allotted, but may still be processing the request (get flowschemas.flowcontrol.apiserver.k8s.io exempt), will retry later
Jan  8 12:15:44 localhost k3s[439]: E0108 12:15:44.760922     439 repair.go:118] unable to refresh the service IP block: the server was unable to return a response in the time allotted, but may still be processing the request (get services)
Jan  8 12:15:44 localhost k3s[439]: E0108 12:15:44.764442     439 repair.go:75] unable to refresh the port block: the server was unable to return a response in the time allotted, but may still be processing the request (get services)
```
`

**Comment by nicklasfrahm** [comment]
I am having a similar issue in a non-HA setup. I attached some logs from `journalctl -fu k3s`:

```
Jan 09 13:13:20 odroidhc2-0 k3s[29844]: E0109 13:13:20.671577   29844 cronjob_controller.go:123] Failed to extract job list: the server was unable to return a response in the time allotted, but may still be processing the request (get jobs.batch)
Jan 09 13:13:22 odroidhc2-0 k3s[29844]: E0109 13:13:22.136719   29844 kubelet_node_status.go:447] Error updating node status, will retry: error getting node "odroidhc2-0": Get "https://127.0.0.1:6443/api/v1/nodes/odroidhc2-0?timeout=10s": context deadline exceeded
Jan 09 13:13:22 odroidhc2-0 k3s[29844]: E0109 13:13:22.526832   29844 memcache.go:179] couldn't get current server API group list: Get "https://127.0.0.1:6444/api?timeout=32s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
Jan 09 13:13:22 odroidhc2-0 k3s[29844]: I0109 13:13:22.528014   29844 event.go:291] "Event occurred" object="traefik/traefik" kind="HorizontalPodAutoscaler" apiVersion="autoscaling/v2beta2" type="Warning" reason="FailedGetScale" message="Get \"https://127.0.0.1:6444/api?timeout=32s\": net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
Jan 09 13:13:32 odroidhc2-0 k3s[29844]: E0109 13:13:32.147250   29844 kubelet_node_status.go:447] Error updating node status, will retry: error getting node "odroidhc2-0": Get "https://127.0.0.1:6443/api/v1/nodes/odroidhc2-0?timeout=10s": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Jan 09 13:13:32 odroidhc2-0 k3s[29844]: time="2021-01-09T13:13:32.629035626Z" level=info msg="Cluster-Http-Server 2021/01/09 13:13:32 http: TLS handshake error from 172.16.0.1:53422: EOF"
Jan 09 13:13:33 odroidhc2-0 k3s[29844]: E0109 13:13:33.902991   29844 node_lifecycle_controller.go:804] Failed while getting a Node to retry updating node health. Probably Node odroidhc2-0 was deleted.
Jan 09 13:13:33 odroidhc2-0 k3s[29844]: E0109 13:13:33.903095   29844 node_lifecycle_controller.go:809] Update health of Node '' from Controller error: the server was unable to return a response in the time allotted, but may still be processing the request (get nodes odroidhc2-0). Skipping - no pods will be evicted.
```

Not sure if this is relevant, but when I restart the node via `sudo systectl restart k3s`, theres is a window of a couple of minutes, where the API server is replying to requests.

**Comment by i5Js** [comment]
Yes, indeed, the service is up and down all the time. My main concern also is it seems there is no a way restore an existing installation from a external database.

**Comment by i5Js** [comment]
How can be possible, if I've only updated two nodes:

```
Jan  9 14:40:33 localhost k3s[14133]: time="2021-01-09T14:40:33.229728732+01:00" level=error msg="Failed to retrieve agent config: https://127.0.0.1:6443/v1-k3s/serving-kubelet.crt: 500 Internal Server Error"
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.342569   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.344083   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.388542   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.388546   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.479741   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.480156   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.480996   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.479741   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.496797   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.498393   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.499774   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.501083   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.502404   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.503846   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.505428   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506600   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.505918   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506076   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.508618   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506216   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506342   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506076   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506478   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
Jan  9 14:40:33 localhost k3s[14133]: E0109 14:40:33.506472   14133 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, Token has been invalidated]
```

**Comment by nicklasfrahm** [comment]
I am not sure about that. Did you try the following on all:
1. Run `sudo systemctl stop k3s` on all nodes
1. Run `BINARY=k3s-armhf sudo curl -SL https://github.com/rancher/k3s/releases/download/v1.19.5+k3s2/$BINARY -o /usr/local/bin/k3s`, while making sure that the `BINARY` variable is set to:
    - `k3s-armhf` for `$(uname -m) == "armv7l"`
    - `k3s-arm64` for `$(uname -m) == "aarch64"`
    - `k3s` for `$(uname -m) == "x86_64"`
1. Run `sudo systemctl restart k3s` on the control plane nodes
1. Wait until all control plane nodes show up as ready and with the version `v1.19.5+k3s2` by running `watch kubectl get nodes`
1. Run `sudo systemctl restart k3s` on all worker nodes

**Comment by i5Js** [comment]
So, downgrade from 1.20 to 1.19 right?

**Comment by nicklasfrahm** [comment]
Sorry if the instructions are a little verbose and tedious, I am running my deployments via ansible and didn't want to take chances to mess up my cluster even more.

**Comment by i5Js** [comment]
No worries, the steps are clear, I'm just finishing in the Master 1 and Worker 1. I'll let you know soon.

**Comment by i5Js** [comment]
No lucky, it still flapping from down to up. Same kind of issues.

**Comment by nicklasfrahm** [comment]
Do you have a backup of your MySQL database?

**Comment by i5Js** [comment]
No, I'm afraid not... 
I'm going to create a new clusters with VMs this time, try to upgrade them, and see if there are no issues. I'm a bit tired to recreate the cluster one time and another in my raspberrys

**Comment by nicklasfrahm** [comment]
I get that, maybe consider deploying your cluster via CI / CD on top of the manually provisioned Pis. That is what works for me, but I am also not running an HA setup. If you do it via CI / CD, you can create a fairly simple backup by creating an automated MySQL dump and zip it. If you then run it in something like GitLab CI, you can upload that as a **protected** artifact and boom you have automatic backups of your cluster state.

**Comment by i5Js** [comment]
Awesome! I'm going to check it, I see you have also Armbian, like me.

Sorry for ask, but what means CI / CD?

**Comment by nicklasfrahm** [comment]
No reason to be sorry, I tend to get lost in technical jibberish. Sorry for that, it is not very inclusive. CI (continuous integration) "is the practice of merging all developers' working copies to a shared mainline several times a day", whereas CD (continous delivery ) is an approach to "produce software in short cycles, ensuring that the software can be reliably released at any time" according to Wikipedia. This is often done by automating your builds, tests and deployments via for example GitHub actions or GitLab CI / CD. 😄 🚀

**Comment by i5Js** [comment]
Well I've deployed a new cluster with VMs, all Debian 10 2 Masters and 2 Workers. I've upgraded them, remove one node, rebooted another, and everything worked as expected. There are only two differences with my pi's installations: The masters have not any taint configuration, and the kernel is 4 instead 5....

**Comment by brandond** [comment]
There is a known issue with arm on v1.20 caused by a golang compiler issue. v1.20.2 will include a workaround: https://github.com/kubernetes/kubernetes/issues/97685

**Comment by i5Js** [comment]
Hello @brandond Thanks the info. Since I'm unable to restore my actual cluster due to certificate issue, even after downgraded the masters to 1.19, I'm going to reinstall it to 1.19 and wait till future versions.

**Comment by i5Js** [comment]
Finally I got everything working, I've uninstalled the master nodes, and installed again, with a new token, then delete one worker node and install it again, and so one with all of them. What a nightmare!

**Comment by i5Js** [comment]
I’ve added a new master node, and same issue, bad certificate in all the nodes again....

**Comment by i5Js** [comment]
The issue comes when I add a new master node:
```
export K3S_KUBECONFIG_MODE="644"
export INSTALL_K3S_EXEC=" --tls-san name --no-deploy servicelb --disable traefik --node-taint k3s-controlplane=true:NoSchedule --datastore-endpoint mysql://user:password@tcp(server:3306)/database"
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION={v1.19.5+k3s2} sh -
```
The log after installing:
```
Jan 11 12:42:14 localhost k3s[4755]: E0111 12:42:14.612829    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:14 localhost k3s[4755]: E0111 12:42:14.753393    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:14 localhost k3s[4755]: E0111 12:42:14.756217    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:15 localhost k3s[4755]: E0111 12:42:15.626156    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:15 localhost k3s[4755]: E0111 12:42:15.767258    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:15 localhost k3s[4755]: E0111 12:42:15.767268    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:16 localhost k3s[4755]: E0111 12:42:16.647785    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:16 localhost k3s[4755]: E0111 12:42:16.780543    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:16 localhost k3s[4755]: E0111 12:42:16.780538    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:17 localhost k3s[4755]: E0111 12:42:17.665438    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:17 localhost k3s[4755]: E0111 12:42:17.793688    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:17 localhost k3s[4755]: E0111 12:42:17.793693    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:18 localhost k3s[4755]: E0111 12:42:18.679277    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:18 localhost k3s[4755]: E0111 12:42:18.808744    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
Jan 11 12:42:18 localhost k3s[4755]: E0111 12:42:18.808792    4755 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose: error in cryptographic primitive]
```

I'm out of ideas....

**Comment by rhuss** [comment]
FYI, I have the very same issue after upgrading to v1.20.0+k3s2 : Timeouts and errors because of an incredible highly load on the master (Pi4, 4GB), so that the cluster is unresponsive.

Downgrading to v1.19.5 didn't help as I the binary crashes more or less immediately. 

Is there any other option for a downgrade except of installing everything afresh ?

**Comment by i5Js** [comment]
Those steps, from @nicklasfrahm : 

1. Run sudo systemctl stop k3s on all nodes
2. Run BINARY=k3s-armhf sudo curl -SL https://github.com/rancher/k3s/releases/download/v1.19.5+k3s2/$BINARY -o /usr/local/bin/k3s, while making sure that the BINARY variable is set to:
3. k3s-armhf for $(uname -m) == "armv7l"
4. k3s-arm64 for $(uname -m) == "aarch64"
5. k3s for $(uname -m) == "x86_64"
6. Run sudo systemctl restart k3s on the control plane nodes
7. Wait until all control plane nodes show up as ready and with the version v1.19.5+k3s2 by running watch kubectl get nodes
8. Run sudo systemctl restart k3s on all worker nodes

I'm pretty tired of try to save this cluster. I'm going to reimage all the rasp, and install it from scratch and no more updates till 1.20 is stable and tested...  it has been more than three days "fighting" with it.

**Comment by rhuss** [comment]
yeah, tried that but the downloaded k3s binary crashes for me with a huge stack trace/thread dump ;-( Trying now with an older 1.19 build (I'm using armv7)

**Comment by brandond** [comment]
@rhuss @i5Js please see my comment above. The extremely high load is caused by a bug in the golang compiler, triggered by code that was added to Kubernetes 1.20. Your choices are to either wait for 1.20.2 later this week, or downgrade to 1.19.5.

**Comment by rhuss** [comment]
yeah, have seen that (and thanks for the heads up !). I'm downgraded now to v1.19.4+k3s1 (not sure why v1.19.5 crashed for me, could have been a flake, too)

Seems to work stable now again. Thanks for your help, waiting now for 1.20.2 :-) (happy to try this out then again)

**Comment by i5Js** [comment]
@brandond, Thanks I saw it and my cluster is in v1.19.5+k3s2 but it doesn't work either... multiples issues about token, certificates and credentials... to far to be stable :)

**Comment by i5Js** [comment]
I've downgraded successfully to V1.19.5. I've only authentications issues in coredns pod:
```
E0114 06:48:25.582249       1 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.4/tools/cache/reflector.go:105: Failed to list *v1.Service: Unauthorized
.:53
[INFO] plugin/reload: Running configuration MD5 = 4665410bf21c8b272fcfd562c482cb82
CoreDNS-1.6.9
linux/arm64, go1.14.1, 1766568
E0114 06:48:26.434807       1 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.4/tools/cache/reflector.go:105: Failed to list *v1.Namespace: Unauthorized
```

Also the masters shows those errors as well:

```
Jan 14 07:52:10 noldork3sM1 k3s[642]: E0114 07:52:10.059741     642 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose
Jan 14 07:52:10 noldork3sM1 k3s[642]: E0114 07:52:10.513351     642 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose
Jan 14 07:52:11 noldork3sM1 k3s[642]: E0114 07:52:11.043335     642 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose
Jan 14 07:52:11 noldork3sM1 k3s[642]: E0114 07:52:11.079320     642 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose
Jan 14 07:52:11 noldork3sM1 k3s[642]: E0114 07:52:11.529060     642 authentication.go:53] Unable to authenticate the request due to an error: [invalid bearer token, square/go-jose
```

Any tips?

**Comment by brandond** [comment]
That usually indicates that there is something running with invalid (revoked or otherwise unrecognized) service account token. Have all the pods been restarted since downgrading?

**Comment by i5Js** [comment]
Hi @brandond many thanks for your answer. No, I didn't. I've killed several times coredns pod, without success, which ones recommends?

**Comment by brandond** [comment]
Cordoning and draining all your nodes, restarting them, and then uncordoning, would be a good way to make sure everything gets flushed out.

**Comment by i5Js** [comment]
Cool, I know with `kubectl drain node` but what about "cordoning" I'll use google, but, if you can explain it to me I would appreciate it.

**Comment by brandond** [comment]
https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service

**Comment by i5Js** [comment]
Ok, I'm "draining" the workers nodes, should I do the same with masters?

**Comment by i5Js** [comment]
I've drained and uncordoned all the workers, and the second master, now the service can't start:

```
Jan 14 10:13:05 localhost k3s[912]: E0114 10:13:05.426801     912 node.go:125] Failed to retrieve node info: nodes "noldork3sm2" is forbidden: User "system:kube-proxy" cannot get resource "nodes" in API group "" at the cluster scope
Jan 14 10:13:06 localhost k3s[912]: time="2021-01-14T10:13:06.337147938+01:00" level=warning msg="Unable to watch for tunnel endpoints: unknown (get endpoints)"
Jan 14 10:13:09 localhost k3s[912]: time="2021-01-14T10:13:09.629652547+01:00" level=fatal msg="networkpolicies.networking.k8s.io is forbidden: User \"system:k3s-controller\" cannot list resource \"networkpolicies\" in API group \"networking.k8s.io\" at the cluster scope"
Jan 14 10:13:09 localhost systemd[1]: k3s.service: Main process exited, code=exited, status=1/FAILURE
Jan 14 10:13:09 localhost systemd[1]: k3s.service: Failed with result 'exit-code'.
Jan 14 10:13:09 localhost systemd[1]: Failed to start Lightweight Kubernetes.
Jan 14 10:13:14 localhost systemd[1]: k3s.service: Service RestartSec=5s expired, scheduling restart.
Jan 14 10:13:14 localhost systemd[1]: k3s.service: Scheduled restart job, restart counter is at 7.
Jan 14 10:13:14 localhost systemd[1]: Stopped Lightweight Kubernetes.
```

**Comment by i5Js** [comment]
All nodes drained, an uncordoned, no luck, same issues :(

**Comment by nicklasfrahm** [comment]
@i5Js I am not exactly a Kubernetes expert, but as @brandond said it could be something with the service account tokens. I am not sure if this might mess with stuff, but you could try to delete the service account token of the coredns pod and recreate it.

The service account token is stored in a secret `kubectl -n kube-system get secrets | grep "coredns"`. I don't know the commands from the top of my head, but you might find stuff online. Later during the day I might have time to write up the commands.

**Comment by i5Js** [comment]
@nicklasfrahm You're my fucking hero :D, as you recommended:
-  I removed the secret : `kubectl delete secret coredns-token-cj77r -n kube-system`
- Delete the pod: `kubectl delete pod coredns-66c464876b-25kl9 -n kube-system`
- Everything working flawless....

Many thanks !

**Comment by cipy** [comment]
hi @i5Js what are you using as k3s slaves? thanks,

**Comment by i5Js** [comment]
Hi @cipy I've two raspberries 2 as masters and as workers I have three raspberries 3 and 2 Nanopiv4

**Comment by rthamrin** [comment]
have you tried to renew the certificate https://www.youtube.com/watch?v=gpGCfj3DqoI

**Comment by cawoodm** [comment]
I am having the same problem - all I did was stop one master node A and try to run it with additional `kube-apiserver-arg` arguments. When this failed, I restarted the node A with the original configuration and suddenly the other master nodes B and C failed and could not communicate with node A citing `invalid bearer token, service account token has been invalidated`.

A pity there is no explanation of what this error means, how it can arise and how to resolve it. What is the token? Where is it stored? How can it become lost?
